parlai/agents/transformer/module_extensions.py

```diff
@@ -0,0 +1,195 @@
+            self.noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.eps, high=self.eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        if self.use_r3f and not self.noise_encoder and not self.noise_decoder:
+            raise RuntimeError(
+                "R3FContext: Noise must be added to at least one of the encoder or decoder."
+            )
+        self.had_first_pass = (
+            False
+        )  # necessary because nn.Embedding for encoder/decoder on some models are the same. Use this to only make any relevant deep copies once.
+
+    def _calculate_symm_kl(self, noised_logits, input_logits):
+        return (
+            F.kl_div(
+                F.log_softmax(noised_logits, dim=-1, dtype=torch.float32),
+                F.softmax(input_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+            + F.kl_div(
+                F.log_softmax(input_logits, dim=-1, dtype=torch.float32),
+                F.softmax(noised_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+        ) / noised_logits.size(0)
+
+    def forward_pass_with_r3f(self, model, *args, **kwargs):
+        """
+        'model' argument here is the `nn.Module` normally responsible for the full
+        forward pass.
+        """
+        if not self.use_r3f:
+            raise RuntimeError(
+                "Trying to use an R3F loss when this has not been allowed."
+            )
+        regular_output = model(*args, **kwargs)
+        with R3FNoiseEmbeddingContextManager(self, model) as model:
+            noised_output = model(*args, **kwargs)
+        symm_kl = self._calculate_symm_kl(noised_output[0], regular_output[0])
+        return regular_output, self.r3f_lambda * symm_kl
+
+
+class R3FNoisedEmbedding(nn.Module):
+    def __init__(self, noise_sampler):
+        super(R3FNoisedEmbedding, self).__init__()
+        self.noise_sampler = noise_sampler
+
+    def __call__(self, module, input, output):
+        noise = self.noise_sampler.sample(sample_shape=output.shape).to(output)
+        return output + noise
+
+
+class R3FNoiseEmbeddingContextManager(AbstractContextManager):
+    """
+    This class finds the first embedding(s) under the encoder/decoder, depending on what
+    is selected, and adds R3F noise to their output.
+
+    Note that this does not discrinate between the dictionary embedding lookup +
+    position embedding classes. However, seeing as how addition is commutative, this
+    should not be an issue.
+    """
+
+    def __init__(self, context, module):
+        self.encoder_hook = None
+        self.decoder_hook = None
+        self.context = context
+        for name, layer in module.named_modules():
```

Comment:
At this point we have two searches into named modules (one is named_children and one is this one). I think we can simplify and use the full name to match the pattern.

Reply:
D'oh, yeah, that would make it easier.